### PR TITLE
Improve two print online wf commands

### PR DESCRIPTION
**Pull request by CBiasuzzi**
print-online-workflow-metrics and print-online-workflow-performances: use table-formatter, adds params output-format and output-file
ref. issues #179  #180

### CODE REVIEW

File: online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java [hunk 1]
```diff
@@ -250,7 +252,7 @@ private void storeStepMetadata(String workflowId, String stateId, OnlineStep ste
 
     private Map<String, String> getMetrics(String workflowId, String mapName) {
         if (isWorkflowStored(workflowId)) {
-            HashMap<String, String> metrics = new HashMap<String, String>();
+            TreeMap<String, String> metrics = new TreeMap<>();
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java]: Map<String, String> metrics = new TreeMap<>() is better if you don't use a specific implementation method.

File: online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java [hunk 2]
```diff
@@ -289,35 +289,31 @@ public String getCsvMetrics(String workflowId, OnlineStep step) {
...
+                            .sorted(Comparator.comparing(Integer::valueOf))
+                            .forEach(stateId -> {
+                                retTable.add(getStoredMapValues(wfMVStore, stateId, step, stepParamsMap.keySet().size(), paramsIndexes));
+                            });
                 }
-            } catch (IOException e) {
+            } catch (Exception e) {
                 String errorMessage = "error getting cvs data for step " + step.name() + " and wf id " + workflowId;
                 LOGGER.error(errorMessage);
                 throw new RuntimeException(errorMessage);
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java]: I think it's better to keep the cause of the exception: maybe you should use "throw new RuntimeException(errorMessage, e);"

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java [hunk 3]
```diff
@@ -15,114 +19,150 @@
...
+                    String[] headers = new String[metrics.keySet().size() + 1];
+                    headers[0] = "state";
+                    int i = 1;
+                    for (String parameter : metrics.keySet()) {
+                        headers[i] = parameter;
+                        i++;
+                    }
+                    LinkedList<Column> columns = Arrays.stream(headers)
+                            .map(Column::new)
+                            .collect(Collectors.toCollection(LinkedList::new));
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: is it necessary to specify the list implementation used? Could it be replaced by Collectors::toList?

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java [hunk 4]
```diff
@@ -15,114 +19,150 @@
...
+                    try (TableFormatter formatter = PrintOnlineWorkflowUtils.createFormatter(tableFormatterConfig, outputFormat, outputFile,
+                            TABLE_TITLE, columns.toArray(new Column[0]))) {
+                        formatter.writeCell(stateId);
+                        columns.stream().skip(1).forEach(x -> {
+                            try {
+                                formatter.writeCell(metrics.get(x.getName()));
+                            } catch (IOException e) {
+                                throw new RuntimeException(e);
+                            }
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: Is it necessary to wrap the IOException into a RuntimeException? If not these lines could be simpler.

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java [hunk 5]
```diff
@@ -15,114 +19,150 @@
...
+                }
+            } else {
+                List<String[]> metricsTableList = onlinedb.getAllMetrics(workflowId, step);
+                if (metricsTableList.isEmpty()) {
+                    System.err.println("\nNo metrics for workflow " + workflowId + " and step " + step.name());
+                } else {
+                    //first row is metrics' header
+                    LinkedList<Column> columns = Arrays.stream(metricsTableList.get(0))
+                            .map(Column::new)
+                            .collect(Collectors.toCollection(LinkedList::new));
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: LinkedList vs Collectors::toList (see above)

File: online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java (hunk 1 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java]: I wanted to stress the fact that the structure is an ordered map. I find it more readable, this way ...

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 4 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: AFAIK streams/lambda style limits checked exceptions usage: quite clumsy but here needed, I guess

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 3 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: what I wanted to convey to the 'reader' is that the list of columns is a sorted list, built in the same order as the attributes in the header from the stream

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 5 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: what I wanted to convey to the 'reader' is that the list of columns is a sorted list, built in the same order as the attributes in the header from the stream

File: online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java (hunk 2 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java]: You are right.

File: online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java (hunk 1 above)
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/db/OnlineDbMVStore.java]: Ok

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 3 above)
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: A list is sorted by insertion order. If you don't need a "doubly-linked list implementation of the List and Deque interfaces", please replace by Collectors::toList.

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 4 above)
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: Ok

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 5 above)
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: See my previous comment

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 3 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: Ok

File: online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java (hunk 5 above)
Comment (CBiasuzzi) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/PrintOnlineWorkflowMetricsTool.java]: Ok

**Pull request merged**